### ISSUE TITLE
Bug: [no-restricted-imports] allowTypeImports not work with regex pattern

### ISSUE THREAD

**Issue by indooorsman** [typescript-eslint/typescript-eslint#10692]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=PTAEAEBcGcFoEsDmA7A9gJwKYCh4FsAHDSUSATwM1AG9QBDUAX1ADN1U9QBycaAY1SVgdLgG4gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uhXyrPyIAJrUoBbYh2j5UGANrhsORNGhTIAGkyKlOYgEN8g6E1lg5WXUpiIA5ogAe6KLmRkOpVgDoAVJp1WUPrw8BwA7gAqJIgAkhJSMs58sIgBSgC%2BALppYOkB2RB56UA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
import type { a } from '@scope/a';
```
See <https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=PTAEAEBcGcFoEsDmA7A9gJwKYCh4FsAHDSUSATwM1AG9QBDUAX1ADN1U9QBycaAY1SVgdLgG4gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uhXyrPyIAJrUoBbYh2j5UGANrhsORNGhTIAGkyKlOYgEN8g6E1lg5WXUpiIA5ogAe6KLmRkOpVgDoAVJp1WUPrw8BwA7gAqJIgAkhJSMs58sIgBSgC%2BALppYOkB2RB56UA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false>

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  {
    "rules": {
      "@typescript-eslint/no-restricted-imports": [
        "error", {
          "patterns": [{
            "regex": "@scope/.*",
            "allowTypeImports": true
          }]
        }
      ]
    }
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Should not report error.

### Actual Result

Report error: 
```
'@scope/a' import is restricted from being used by a pattern.ESLint[@typescript-eslint/no-restricted-imports](https://typescript-eslint.io/rules/no-restricted-imports)
```

<img width="580" alt="A screenshot of the error" src="https://github.com/user-attachments/assets/e2d90b6d-907f-485a-b71f-da325738f9e1" />

**Comment by yeonjuan** [comment]
The regex option was added in v9.7.0 in the upstream. 
- https://github.com/eslint/eslint/pull/18622
- https://github.com/eslint/eslint/releases/tag/v9.7.0

`typescript-eslint/no-restricted-imports` seems to need an update to support the regex option.
